output/pixel_pusher.c

```diff
@@ -0,0 +1,272 @@
+    }
+
+    if (pp_init_out() < 0) {
+        return -1;
+    }
+
+    return pp_add_grids();
+}
+
+void output_pp_term() {
+    INFO("Terminating PixelPusher");
+
+    if (pp_info) {
+        free(pp_info);
+        pp_info = NULL;
+    }
+
+    for (size_t i = 0; i < n_grid_devices; i++) {
+        struct output_device base = grid_devices[i].base;
+
+        free(base.pixels.xs);
+        free(base.pixels.ys);
+        free(base.pixels.colors);
+
+        if (base.prev != NULL)
+            base.prev->next = base.next;
+        if (base.next != NULL)
+            base.next->prev = base.prev;
+    }
+    free(grid_devices);
+    grid_devices = NULL;
+    n_grid_devices = 0;
+
+
+    if (out_fd) {
+        close(out_fd);
+        out_fd = -1;
+    }
+
+    if (out_addr) {
+        free(out_addr);
+        out_addr = NULL;
+    }
+
+    if (out_packet) {
+        free(out_packet);
+        out_packet = NULL;
+    }
+}
+
+int output_pp_do_frame() {
+    seq_num++;
+
+    ((uint32_t *) out_packet)[0] = seq_num;
+    size_t out_packet_idx = 4;
+
+    for (size_t i = 0; i < n_grid_devices; i++) {
+        struct pp_device * device = &grid_devices[i];
+
+        // Make sure it was successfully initialized
+        if (!device->base.active) continue;
+
+        // Copy the data to send into a buffer - sadly SDL_Color is rgba
+        // so we can't just send a header and it using sendmsg
+        out_packet[out_packet_idx++] = device->strip_num;
+
+        // Snake: The PixelPusher has linear strips arranged into a grid
+        // by going "back and forth".
+        for (int j = 0; j < device->height; j++) {
+            for (int k = 0; k < device->width; k++) {
+                int idx = (j % 2 ? device->width - 1 - k: k)*(device->height) + j;
+
+                SDL_Color color = device->base.pixels.colors[idx];
+                double alpha = color.a / 255.0;
+                out_packet[out_packet_idx++] = color.r * alpha;
+                out_packet[out_packet_idx++] = color.g * alpha;
+                out_packet[out_packet_idx++] = color.b * alpha;
+            }
+        }
+
+        // Send it: the PixelPusher can take 2 grids per packet
+        if ((i % 2) || (i == n_grid_devices - 1)) {
+            printf("Sending packet with i=%zu\n", i);
+
+            size_t sent_bytes = sendto(out_fd, out_packet, out_packet_idx, 0, (struct sockaddr *) out_addr, sizeof *out_addr);
+            if (sent_bytes < out_packet_idx) {
+                PERROR("Error sending PixelPusher data");
+                return -1;
+            }
+
+            // The PixelPusher doesn't have a very large Ethernet buffer, and UDP doesn't resend things,
+            // so if we don't give it some time to process it'll just drop any more incoming packets. The
+            // symptom of this is only some of the grids will respond and the others will stay dark or
+            // flicker; if this happens increase packet_interval.
+            nanosleep(&packet_interval, NULL);
```

Comment:
We could probably skip this on the last packet of the frame; if I'm reading it right output.c has a SDL_Delay to time the output, so this is just extra burn we can't skip if we get behind.